sound/soc/sof/ipc4-topology.c

```diff
@@ -815,6 +817,74 @@ static void sof_ipc4_widget_free_comp_mixer(struct snd_sof_widget *swidget)
 	swidget->private = NULL;
 }
 
+/*
+ * Add the process modules support. The process modules can be EQFIR, EQIIR,
+ * KEYWORD_DETECT, KPB (KeyPhrase Buffer Manager), CHAN_SELECTOR, SMART_AMP
+ * and etc.
+ */
+static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_process *process;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_control *scontrol;
+	struct sof_ipc4_control_data *control_data;
+	void *cfg;
+	int cfg_size;
+	int ret;
+
+	process = kzalloc(sizeof(*process), GFP_KERNEL);
+	if (!process)
+		return -ENOMEM;
+
+	swidget->private = process;
+
+	ret = sof_ipc4_get_audio_fmt(scomp, swidget, &process->available_fmt, true);
+	if (ret)
+		goto err;
+
+	cfg_size = sizeof(struct sof_ipc4_base_module_cfg) +
+		swidget->output_cnt * sizeof(struct sof_ipc4_audio_format);
+	/* allocate the cfg for process modules */
+	list_for_each_entry(scontrol, &sdev->kcontrol_list, list)
+		if (scontrol->comp_id == swidget->comp_id) {
+			if (sof_ipc4_ctl_is_initinstance(scontrol)) {
+				control_data = scontrol->ipc_control_data;
+				cfg_size = cfg_size + control_data->data->size - IPC4_MAGIC_LEN;
+				break;
+			}
+		}
+	cfg = kzalloc(cfg_size, GFP_KERNEL);
+	if (!cfg)
+		goto free_available_fmt;
+	process->ipc_config_data = cfg;
+	process->ipc_config_size = cfg_size;
+
+	ret = sof_ipc4_widget_setup_msg(swidget, &process->msg);
+	if (ret)
+		goto free_available_fmt;
+
+	return 0;
+free_available_fmt:
+	sof_ipc4_free_audio_fmt(&process->available_fmt);
+err:
+	kfree(process);
+	swidget->private = NULL;
+	return ret;
+}
+
+static void sof_ipc4_widget_free_comp_process(struct snd_sof_widget *swidget)
+{
+	struct sof_ipc4_process *process = swidget->private;
+
+	if (!process)
+		return;
+
+	sof_ipc4_free_audio_fmt(&process->available_fmt);
```

Comment:
do we need to free process->ipc_config_data also?

Reply:
> do we need to free process->ipc_config_data also?

Yes. Thanks for pointing it.